Add documentation for API rate limiting (#4287)
opensearch-project/documentation-website: _security/configuration/configuration.md

If `challenge` is set to `false` and no `Authorization` header field is set, the Security plugin does not send a `WWW-Authenticate` response back to the client, and authentication fails. You might want to use this setting if you have another challenge `http_authenticator` in your configured authentication domains. One such scenario is when you plan to use basic authentication and OpenID Connect together.

## Rate limiting

| Setting | Description |
| :--- | :--- |

> | `opensearch_security.ui.openid.login.buttonname` |  Display name for the login button. "Log in with single sign-on" by default. |

Reviewer: Is this change part of this PR?

Author: @DarshitChanpura Just in a draft state right now. I'll keep you posted.

Author: @DarshitChanpura I believe this is all set. When you get a chance, could you have a last look? Thanks for all the help with this.

> If `challenge` is set to `false` and no `Authorization` header field is set, the Security plugin does not send a `WWW-Authenticate` response back to the client, and authentication fails. You might want to use this setting if you have another challenge `http_authenticator` in your configured authentication domains. One such scenario is when you plan to use basic authentication and OpenID Connect together.

Reviewer 2: ```suggestion
If `challenge` is set to `false` and no `Authorization` header field is set, the Security plugin does not send a `WWW-Authenticate` response back to the client, and authentication fails. Consider using this setting if you have the challenge `http_authenticator` in your configured authentication domains. One such scenario is when you plan to use basic authentication and OpenID Connect together.
```

Reviewer 2: Suggestion to clarify second sentence: "Consider setting `challenge` to `false` when you are using `http_authenticator... You'd use this setting, for example, when you plan to..."

## API rate limiting

> API rate limiting is typically used to restrict the number of API calls that users can make in a set span of time, and it can thereby help manage the rate of API traffic. For security purposes, rate limiting features have the potential to defend against DoS attacks, or repeated login attempts to gain access through trial and error, by restricting failed login attempts.

Reviewer 2: ```suggestion
API rate limiting is typically used to restrict the number of API calls that users can make in a set span of time, and thereby help manage the rate of API traffic. For security purposes, rate limiting features have the potential to defend against DoS attacks, or repeated login attempts to gain access through trial and error, by restricting failed login attempts.
```

| `block_expiry_seconds` | The duration of time that login remains blocked after a failed login. After this time elapses, login is reset and the username can attempt successful login again. |
| `max_blocked_clients` |  The maximum number of blocked usernames. This limits heap usage to avoid a potential DoS. |
| `max_tracked_clients` | The maximum number of tracked usernames that have failed login. This limits heap usage to avoid a potential DoS. |

### IP address rate limiting

This configuration limits login attempts by IP address. When a login fails, the IP address specific to the machine being used for login is blocked. 

There are two steps for configuring IP address rate limiting. First, set the `challenge` setting to `false` in the `http_authenticator` section of the `config.yml` file.

```yml
http_authenticator:
  type: basic
  challenge: false
```

For more information about this setting, see [HTTP basic authentication](/security/configuration/configuration/#http-basic-authentication).

> Second, configure the IP address rate limiting settings. The following example shows a completed configuration:

Reviewer 3: ```suggestion
Second, configure the IP address's rate limiting settings. The following example shows a completed configuration:
```

Author: Will get an editorial opinion on this. They aren't really settings for the IP address. The settings are for the feature called "IP address rate limiting". 
Maybe I should call it "IP client rate limiting". That was another alternative from the dev (both were supposed to be accurate). Would that sound better, read better, make more sense when reading it? Thanks.

Reviewer 2: Since we're using the naming convention "IP address rate limiting," leave as originally written. The term is a compound noun. Alternatively, you could use the setting name: "Second, configure the `ip_rate_limiting` setting."
